rfd/0094-kubernetes-node-joining.md

```diff
@@ -0,0 +1,323 @@
+Kubernetes-issued token and "join token" refers to a Teleport join token of
+type `kube`.
+
+If this RFD is accepted, the teleport documentation should reflect this nuance
+and using the term token alone should be avoided.
+
+The feasibility of this approach has been tested against Multiple Kubernetes
+setups, including:
+- Kubernetes 1.18 on Minikube
+- Kubernetes 1.20 on EKS
+- Kubernetes 1.22 on GKE
+
+The proof of concept code can be found
+[here](https://github.com/hugoShaka/kube-JWT-playground).
+
+#### User flow
+
+- A teleport cluster is running in Kubernetes.
+- An operator (human or automated) wants to deploy a new set of Teleport
+  proxies in the same cluster.
+- The operator creates a new `kube` join token, allowing the kubernetes service
+  account `new-proxies`
+- The operator creates a Kubernetes service account `new-proxies`
+- The operator creates its new pods (with a Deployment, a StatefulSet, or even
+  a DaemonSet) using the `new-proxies` SA
+- The Teleport proxies uses the mounted Kubernetes token to join the existing
+  Teleport cluster
+- The Teleport cluster validates the token against Kubernetes using the
+  `TokenReview` API
+- The proxies are part of the cluster
+
+#### Kubernetes tokens
+
+All Kubernetes tokens contain the service account name and its groups.
+
+Versions 1.20 and above ([bound
+tokens](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/1205-bound-service-account-tokens/README.md#boundserviceaccounttokenvolume-1)):
+- tokens are signed per pod
+- tokens are automatically renewed by the kubelet
+- token TTL is configurable in the `PodSpec`
+- tokens are revoked when the pod is deleted
+
+Versions prior 1.20:
+- tokens are signed per service-account
+- tokens are long-lived
+- tokens are revoked when the service account is deleted
+
+Tokens are JWTs, Kubernetes JWKS can be manually fetched and JWT validation can
+be done offline.  Kubernetes also exposes a `TokenReview` API to validate
+tokens. This API performs more checks, like revocation, UUID checking pod and
+service account, ... This API is the sanest way to validate Kubernetes token.
+
+#### Teleport join tokens
+
+The teleport resource would look like:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: proxy-token
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  roles: ["Proxy"]
+  join_method: kube
+  allow:
+  - service_account: "my-namespace:my-service-account"
+```
+
+The `allow[*].service_account`check should match exactly the `User` field of
+the Kubernetes token stripped of the `system:serviceaccounts:` prefix. Checking
+if the `system:serviceaccounts` prefix is present ensures no Kubernetes user
+tokens can be used to join the Teleport cluster.
+
+Supporting this token kind does not require changes to the
+`RegisterUsingTokenRequest` type, the Kubernetes token should be passed through
+the `Token` field and the Teleport token name should be passed in the `TokenID`
```

Comment:
```suggestion
the `IDToken` field and the Teleport token name should be passed in the `Token`
```

I believe you currently have these reversed, typically `Token` has been used to specify the ProvisionToken that is being used for joining.